libcib, libpe_status: Replace snprintf() with GString

nrwahl2 opened the pull request:
Skip the output functions in `bundle.c` for now: `pe__bundle_replica_output_html()`, `pe__bundle_replica_output_text()`, and `print_bundle_replica()`. Since we ultimately pass to `fprintf()`, it might make sense to use `LINE_MAX` as a size limit... not entirely sure. We can come back to those later if we want.

Some performance test results after each successive change, relative to the commits in https://github.com/ClusterLabs/pacemaker/pull/2857:
```
# tools/pcmk_simtimes -S percent /tmp/before_1000.txt /tmp/after_unpack_1000.txt | tail -1
Averages: +0.01s  +0.50%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/before_1000.txt /tmp/after_unpack_1000.txt | tail -1
Averages: +0.08s  +0.52%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/before_1000.txt /tmp/after_unpack_1000.txt | tail -1
Averages: +0.04s  -0.16%

# tools/pcmk_simtimes -S percent /tmp/before_1000.txt /tmp/after_bundle_1000.txt | tail -1
Averages: +0.02s  +0.58%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/before_1000.txt /tmp/after_bundle_1000.txt | tail -1
Averages: +0.15s  +0.63%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/before_1000.txt /tmp/after_bundle_1000.txt | tail -1
<empty>

# tools/pcmk_simtimes -S percent /tmp/before_1000.txt /tmp/after_pe_actions_1000.txt | tail -1
Averages: +0.02s  +0.65%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/before_1000.txt /tmp/after_pe_actions_1000.txt | tail -1
Averages: +0.21s  +0.37%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/before_1000.txt /tmp/after_pe_actions_1000.txt | tail -1
Averages: -0.13s  -4.69%

# tools/pcmk_simtimes -S percent /tmp/before_1000.txt /tmp/after_cib_attrs_1000.txt | tail -1
Averages: +0.01s  +0.58%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/before_1000.txt /tmp/after_cib_attrs_1000.txt | tail -1
Averages: +0.08s  +0.47%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/before_1000.txt /tmp/after_cib_attrs_1000.txt | tail -1
Averages: -0.14s  -5.34%
```

nrwahl2 commented:
Added another commit on top, to use GString in `pcmk__add_separated_word()`.
```
# tools/pcmk_simtimes -S percent /tmp/before_1000.txt /tmp/after_pcmk__add_word_1000.txt | tail -1
Averages: -0.06s  -1.09%
# tools/pcmk_simtimes -s 0.1 -S percent /tmp/before_1000.txt /tmp/after_pcmk__add_word_1000.txt | tail -1
Averages: -0.37s  -1.97%
# tools/pcmk_simtimes -s 0.1 -p 3 -S percent /tmp/before_1000.txt /tmp/after_pcmk__add_word_1000.txt | tail -1
Averages: -0.63s  -3.49%
```

nrwahl2 commented on lib/pengine/pe_notif.c:
```diff
@@ -633,13 +625,22 @@ collect_resource_data(pe_resource_t *rsc, bool activity, notify_data_t *n_data)
     }
 }
 
+// For (char *) value
```
We still use `add_notify_env()` in `pe__order_notifs_after_fencing()`, and there's no reason to use a `GString` there. It felt most straightforward to use separate macros for regular strings and `GString`s.

No need for a regular string macro for `add_notify_env_free_gs()` but it seemed right to change the name.

nrwahl2 commented on lib/pengine/clone.c:
```diff
@@ -697,9 +695,9 @@ clone_print(pe_resource_t * rsc, const char *pre_text, long options, void *print
-            free(stopped_list);
-            stopped_list = NULL;
-            stopped_list_len = 0;
+            if (stopped_list != NULL) {
```
I'm a little bit confused about the existing logic. If `!pcmk_is_set(rsc->flags, pe_rsc_unique` and `clone_data->clone_max > active_instances`, then we blow away the `stopped_list` without ever printing it. Then we may overwrite it with something else and print that.

Do we want to print the original `stopped_list` before we create the "Custom stopped list for non-unique clones"? If not, should we avoid building it?

kgaillot replied:
We don't want to print it, the point is that the default stopped list isn't appropriate for anonymous clones when more instances could run. Avoiding building it in the first place would make more sense.

However note that the *print() functions are deprecated and unused. The pcmk__output_t model obsoleted them, but they're implementations of public API, so we need to keep them around for now. For the most part we just ignore them, so really all of this isn't necessary. FYI the next API compatibility break is expected in mid-2024 (which would be the 2.2.0 or 3.0.0 release).

nrwahl2 replied:
That's a long way off. On the bright side, hopefully that's enough time to deprecate everything we want to deprecate before then.

nrwahl2 commented on daemons/controld/controld_execd.c:
```diff
@@ -632,22 +631,27 @@ append_restart_list(lrmd_event_data_t *op, struct ra_metadata_s *metadata,
     crm_xml_add(update, XML_LRM_ATTR_RESTART_DIGEST, digest);
 
-    crm_trace("%s: %s, %s", op->rsc_id, digest, list);
+    if (list != NULL) {
```
Currently assumes `list` is non-null. I guess that's probably a safe assumption

nrwahl2 replied:
> We also don't need to trace it twice, either one would be sufficient.

What exactly are you referring to? I'm about to make the regular append mirror the secure append but it sounds like something else is being asked, since they both have the same number of trace calls... Only differences w.r.t. tracing are that `crm_log_xml_trace()` is called unconditionally here and conditionally in secure, and the digest is traced unconditionally here.

nrwahl2 replied:
Looking at `calculate_operation_digest()`, it calls `calculate_xml_digest_v1()`, which traces its input (`restart`)... so maybe that's what you meant. It looks like the same XML will get traced twice here -- granted, tracing might be enabled in only one location.

nrwahl2 replied:
I think I jumped the gun on this change in the first place. It looks like we do still want to do some things even if there are no reloadable or unique params. For now, I'm gonna revert to the original structure and just add NULL checks where applicable.

kgaillot replied:
> It looks like we do still want to create a digest and add it to the XML even if list is empty.

Ah right, I overlooked that too. I believe its presence triggers some code.

nrwahl2 commented on lib/pacemaker/pcmk_graph_logging.c:
```diff
@@ -169,16 +182,20 @@ static void
-    char *pending = NULL;
+    GString *pending = NULL;
 
     if (!pcmk_is_set(synapse->flags, pcmk__synapse_executed)) {
```
Is this even right? Based on `synapse_state_str()`, a synapse with the executed flag set can be in-flight.

Also, as the coverity error pointed out, we'll currently dereference a null pointer if the executed flag is set (so we skip this block) and then we enter the for-loop.

It seems like we'd want to either:
* enclose everything inside the `if (!pcmk_is_set(synapse->flags, pcmk__synapse_executed))` block (for whatever reason, not sure why it exists) so that `pending` is non-NULL when the for-loop runs;
* assign `"none"` to `pending` if the executed bit is set... which seems wrong, since it could be in-flight; or
* take the `pending = synapse_pending_inputs(graph, synapse)` line out of the if-block so that we run it no matter what.

I lean toward option 3 but there might be a reason things are the way they are.

nrwahl2 replied:
We did lose a null check on `pending` in 5b33ce3; previously we logged `"none"` if `pending == NULL`. That's a trivial enough fix.

However, even before that commit, we ran the equivalent of `synapse_pending_inputs()` only if `synapse->executed` was false. But `synapse_pending_inputs()` considers in-flight actions (based on the `// In-flight or pending` comment), and it sounds like a synapse with `executed` set to true could have in-flight actions... so I'm not sure why we'd want to guard the `synapse_pending_inputs()` call inside `if !pcmk_is_set(synapse->flags, pcmk__synapse_executed))`.

kgaillot replied:
> But it's odd that coverity didn't complain about the original:

free() accepts (and ignores) NULL, but most glib functions don't (the GList functions are an exception since NULL is a valid empty GList)

nrwahl2 replied:
Coverity complains about dereferencing a null pointer here:
```
        log_synapse_action(log_level, synapse,
                           (pcmk__graph_action_t *) lpc->data,
                           (const char *) pending->str);
```
not in the free

kgaillot replied:
> It seems like we'd want to either:

The current code is just a style preference. Some people prefer to have as little indentation as possible. But even better for that purpose would be to do "if (pcmk_is_set(...)) { return; }" at the start. Enclosing the entire thing in "if (!pcmk_is_set(...))" is also a reasonable approach.

nrwahl2 replied:
> However, even before that commit, we ran the equivalent of synapse_pending_inputs() only if synapse->executed was false. But synapse_pending_inputs() considers in-flight actions (based on the // In-flight or pending comment), and it sounds like a synapse with executed set to true could have in-flight actions

I was failing to differentiate between the synapse being in-flight versus its pending actions being in-flight. The if-guard makes more sense now.

nrwahl2 replied:
> Some people prefer to have as little indentation as possible.

I prefer to return/continue/break early to have less indentation myself

kgaillot replied:
> We did lose a null check on pending in https://github.com/ClusterLabs/pacemaker/commit/5b33ce3995ce08d68519c5c329991ef9010fba9c; previously we logged "none" if pending == NULL. That's a trivial enough fix.

synapse_pending_inputs() returns "none" for NULL now

kgaillot replied:
> I guess because the dereference is now explicit in this function instead of in whatever qb function do_crm_log calls

Yeah you're right, either log_synapse_action() should use pcmk__s() on pending_inputs, or pending should be strdup("none") if synapse_pending_inputs() isn't called

nrwahl2 replied:
> synapse_pending_inputs() returns `"none"` for NULL now

Which is great, if `synapse_pending_inputs()` gets called :)

The latest update gets rid of the `strdup("none")`/`g_string_new("none")` altogether, sets a `const char *` to "none" by default in `log_synapse()`, and updates it to point to the pending list if `synapse_pending_inputs()` is called and returns non-`NULL`.

kgaillot reviewed (commented):
looks great

nrwahl2 commented on lib/pengine/bundle.c:
```diff
@@ -1269,14 +1278,16 @@ pe__unpack_bundle(pe_resource_t *rsc, pe_working_set_t *data_set)
         }
-        bundle_data->container_host_options = buffer;
+        bundle_data->container_host_options = strdup((const char *)
+                                                     buffer->str);
```
Good point.
It's questionable whether we need to strdup before assigning to a char * anyway. I've been doing it out of an abundance of caution for memory management (g_free vs. free). But there are existing places in the code where a gstring->str is assigned directly to a char pointer and then g_string_free(gstring, FALSE) is called (freeing the gstring but not gstring->str).

nrwahl2 replied:
As an FYI, since I got curious... Since glib 2.46, `g_malloc()` is guaranteed to use the system `malloc()`, and `g_free()` just calls the system `free()`.

* https://github.com/GNOME/glib/blob/main/glib/gmem.c#L106
* https://github.com/GNOME/glib/blob/main/glib/gmem.c#L226-L231

So in recent versions, it seems like it's not strictly required, although it is a best practice. To be fair, I can't figure out what the `TRACE(GLIB_MEM_FREE(...))` call is doing. `GLIB_MEM_FREE` is not defined anywhere in the glib source code, so it's something more obscure...

Anyway, Pacemaker only requires glib >= 2.42, so this is moot.

nrwahl2 commented on lib/pengine/unpack.c:
```diff
@@ -2591,18 +2591,22 @@ find_lrm_op(const char *resource, const char *op, const char *node, const char *
 
+    CRM_CHECK((resource != NULL) && (op != NULL) && (node != NULL),
+              return NULL);
+
```
I've always seen it more as a way to make future bugs obvious (e.g., if a buggy caller gets added later). Lots of static functions have assertions or checks, and lots of others don't. I don't have a strong preference so any time you'd prefer to skip it to make the function shorter, I'm good with that.

kgaillot replied:
No, it's purely a judgment call. I just wanted to give you some background on when it's more or less important

nrwahl2 commented:
Updated to address review.
* Replaced fall through comment
* Removed `#if 0` sections that contained string append functions (left the ones with the "Consider supporting" TODO)
* Made `pe__bundle_variant_data_t:container_host_options` a `gchar *` to avoid a strdup.
* Reverted some questionable changes (where I had put more things into an `if (list != NULL)` guard) in `append_restart_list()`.

nrwahl2 commented on lib/pengine/clone.c:
```diff
@@ -407,13 +405,14 @@ clone_active(pe_resource_t * rsc, gboolean all)
 static void
-short_print(char *list, const char *prefix, const char *type, const char *suffix, long options, void *print_data)
+short_print(const char *list, const char *prefix, const char *type,
+            const char *suffix, long options, void *print_data)
```
Added a commit on top for this. I think I got them all

nrwahl2 commented:
Updated to fix a bug found in the `pcmk__add_separated_word()` commit in `create_level_registration_xml()` (lack of `NULL` check at the end). Caused fencer regression tests to fail.

kgaillot reviewed (commented):
looking good

nrwahl2 commented on lib/pengine/bundle.c:
```diff
@@ -1287,19 +1281,19 @@ pe__unpack_bundle(pe_resource_t *rsc, pe_working_set_t *data_set)
-        bundle_data->container_host_options = buffer;
+        bundle_data->container_host_options = strdup((const char *)
+                                                     buffer->str);
+        g_string_free(buffer, TRUE);
```
You mean asserting on a `NULL` `strdup()` result, or something else?

nrwahl2 replied:
Oh I see, the `gchar *` conversion

kgaillot replied:
Yep, though now that you mention it, just dropping the options would be bad, so asserting would be good too

nrwahl2 replied:
What did you have in mind? We allocate the `GString` buffer unconditionally (to match the previous `calloc()` call), so worst case scenario `buffer->str` will be the empty string. Then when we free `buffer`, the data in `buffer->str` gets assigned to `bundle_data->container_host_options`.

I did accidentally re-declare `buffer` in the `else` block (will fix momentarily) but it doesn't cause a leak.

nrwahl2 replied:
> accidentally re-declare `buffer

Disregard, I was misreading the diff just now :)

kgaillot replied:
nm, i was looking at an older diff, all good

nrwahl2 commented on lib/pengine/pe_actions.c:
```diff
@@ -1235,20 +1232,20 @@ pe_fence_op(pe_node_t * node, const char *op, bool optional, const char *reason,
             g_hash_table_insert(stonith_op->meta,
                                 strdup(XML_OP_ATTR_DIGESTS_SECURE),
-                                digests_secure);
+                                strdup((const char *) digests_secure->str));
```
Agreed

nrwahl2 commented:
Updated bundle.c and pe_actions.c commits to address review
